rules/prefer-error-cause.js

```diff
@@ -0,0 +1,320 @@
+				targetArguments = thrownErrorDeclarator.right.arguments;
+			}
+
+			errorConstructorLastArgument = peekTop(targetArguments);
+
+			if (!errorConstructorLastArgument) {
+				reportCannotBeFixed(catchBlock, context);
+				return [];
+			}
+
+			result.push({
+				node: errorConstructorLastArgument,
+				statementToFix,
+			});
+		}
+	}
+
+	return result;
+};
+
+const handleCatchBlock = ({context, catchBlock, parameter, throwStatement}) => {
+	const errorArgumentIdentifier = parameter?.name;
+	const throwStatementArgument = throwStatement.argument;
+
+	// Filter rethrowing the error itself.
+	// `try {} catch (err) { throw err; }`
+	// `promise.then(undefined, (err) => { throw err; })`
+	// `promise.catch(err => { throw err; })`
+	if (errorArgumentIdentifier && errorArgumentIdentifier === throwStatementArgument.name) {
+		return;
+	}
+
+	// Report none identifier parameter
+	// `try {} catch ({error}) {}`
+	// `promise.then(undefined, ({error}) => {})`
+	// `promise.catch({error} => {})`
+	if (parameter && parameter.type !== 'Identifier') {
+		context.report({
+			node: catchBlock,
+			messageId: ERROR,
+		});
+		return;
+	}
+
+	const nodesToFixCandidates = getAllNodesToFix({catchBlock, context, throwStatement});
+
+	const nodesToFix = nodesToFixCandidates.filter(({node}) =>
+		!hasCauseProperty(node),
+	);
+
+	for (const nodeToFix of nodesToFix) {
+		context.report({
+			node: catchBlock,
+			messageId: ERROR,
+			fix(fixer) {
+				return fix({
+					fixer,
+					context,
+					catchBlock,
+					statementToFix: nodeToFix.statementToFix,
+					errorConstructorLastArgument: nodeToFix.node,
+					errorArgumentIdentifier,
+				});
+			},
+		});
+	}
+};
+
+const isThenMethod = callExpression =>
+	callExpression?.optional !== true
+	&& callExpression?.arguments?.length === 2
+	&& callExpression?.arguments[0].type !== 'SpreadElement'
+	&& callExpression?.arguments[1].type !== 'SpreadElement'
+	&& callExpression?.callee.type === 'MemberExpression'
+	&& callExpression?.callee.optional !== true
+	&& callExpression?.callee.computed !== true
+	&& callExpression?.callee.property.type === 'Identifier'
+	&& callExpression?.callee.property.name === 'then';
+
+const isCatchMethod = callExpression =>
+	callExpression?.optional !== true
+	&& callExpression?.arguments?.length === 1
+	&& callExpression?.arguments[0].type !== 'SpreadElement'
+	&& callExpression?.callee.type === 'MemberExpression'
+	&& callExpression?.callee.optional !== true
+	&& callExpression?.callee.computed !== true
+	&& callExpression?.callee.property.type === 'Identifier'
+	&& callExpression?.callee.property.name === 'catch';
+
+/** @param {import('eslint').Rule.RuleContext} context */
+const create = context => ({
+	[throwStatementSelector](node) {
+		const catchBlock
+			= getMatchingAncestorOfType(node, 'CatchClause')
+			|| getMatchingAncestorOfType(node, 'CallExpression', parent => isThenMethod(parent) || isCatchMethod(parent));
```

Comment:
I haven't read all the code, but I guess you are searching for the "old" error, but I don't think this is the correct way.

```js
try {} catch (error1) {
   return foo(bar).catch(error2 => {
       // This should be caused by `error2` not `error1`
    throw new Error();
  })
      
  function foo(bar) {
     // Not caused by `error1`
     if (bar < 0)
       throw new Error();
  }
}
```

---

I guess we should look for `CatchClause` or `:function`, 
1. not `CatchClause` first.
2. all function, not only argument of `Promise#{then, catch}()`, and check if the function is a `{then,catch}` argument.

Reply:
I think this is resolved now if I understand correctly.
I replaced existing `getMatchingAncestorOfType` call with `getCatchBlock` for resolving (1).
And I used `esquery` in the `getCatchBlock` for implementing for checking if the function is `{then, catch}`'s argument (2).
And `getCatchBlock` return undefined if the throw statement is in the function to avoid existing wrong fix. (Fixing throw statement in the `function foo` in your example code)